### ISSUE TITLE
please fix it. thanks

### ISSUE THREAD

**Issue by JamesLinus** [OnionBrowser/iOS-OnionBrowser#244]
open multiple web pages. the app will crashes.

**Comment by tladesignz** [comment]
Well... Unfortunately, it's not *that* easy. If it would crash always when opening a few pages, then we would have noticed and fixed it already, because that would have been some obvious error.

So, I will need way more details to be able to help you. Like:

- How many do I need to open?
- Is it always the same number of tabs or does it depend on certain websites?
- What websites should I open to reproduce it?
- What's your device and iOS version?

I intentionally didn't limit the number of tabs, so if you try hard enough, you'll surely crash the app when running out of memory. So, in a sense, it's a feature.

I could limit the number of tabs to save users from themselves, but the real problem is not so much how many tabs there are, but how much memory the open websites consume. There are very lightwight ones, and then there are heavy webapps, which easily do hundreds of megabytes at once. Open Facebook, Twitter and some more of these at the same time, and you surely hit the memory limit.

Come to think of it, we could unload webpages, when we get a memory warning. But, anyway, would be good to figure out what's a real-life scenario!

**Comment by tladesignz** [comment]
Ok, so I found a massive memory leak (uuups), which will be fixed in the next version.

Also, HTML content of background tabs will now be unloaded on iOS memory warnings.

@mtigas: Is there a way to send Tor a signal that we're under memory pressure and it should give up some memory?

**Comment by mtigas** [comment]
> @mtigas: Is there a way to send Tor a signal that we're under memory pressure and it should give up some memory?

I'm not aware of anything off the top of my head. I don't _think_ there is.

**Comment by tladesignz** [comment]
> > @mtigas: Is there a way to send Tor a signal that we're under memory pressure and it should give up some memory?
> 
> I'm not aware of anything off the top of my head. I don't _think_ there is.

See, when I tested with 18 open tabs full of news websites, I reached around 900 MByte of RAM usage. Closing the tabs, it fell back to ~550 MByte. After launch, though, we start with ~150 MByte RAM usage.

When I do a memory graph debug, I see **a lot** libboringssl, libnetwork and CFNetwork  objects and a **huge** amount of Malloc blocks. Also, the analyzer shows a dozen retain cycles, all of which involve a "SecPolicy" object. Hmmmmm.
